qCC/ccCommandLineCommands.cpp

```diff
@@ -1451,93 +1436,41 @@ bool CommandSubsample::process(ccCommandLineInterface& cmd)
 				}
 				else
 				{
-					//found active scalar field
-					ScalarType sfMin = CCCoreLib::NAN_VALUE;
-					ScalarType sfMax = CCCoreLib::NAN_VALUE;
-					if (sf->countValidValues() > 0)
-					{
-						if (!ccScalarField::ValidValue(sfMin) || sfMin > sf->getMin())
-							sfMin = sf->getMin();
-						if (!ccScalarField::ValidValue(sfMax) || sfMax < sf->getMax())
-							sfMax = sf->getMax();
-						if (!ccScalarField::ValidValue(sfMin) || !ccScalarField::ValidValue(sfMax))
-						{
-							//warn the user, don't use 'Use Active SF' and keep going
-							cmd.warning(QObject::tr("\tCan't use 'Use active SF': scalar field '%1' has invalid min/max values.").arg(QString::fromStdString(sf->getName())));
-						}
-						else
-						{
-							//everything validated use acitve SF for modulation
-							//implementation of modParams.a/b values come from MainWindow::doActionSubsample()
-							modParams.enabled = true;
-
-							double deltaSF = static_cast<double>(sfMax) - static_cast<double>(sfMin);
-
-							if (CCCoreLib::GreaterThanEpsilon(deltaSF))
-							{
-								modParams.a = (sfMaxSpacing - sfMinSpacing) / deltaSF;
-								modParams.b = sfMinSpacing - modParams.a * sfMin;
-							}
-							else
-							{
-								modParams.a = 0.0;
-								modParams.b = sfMin;
-							}
-							cmd.print(QObject::tr("\tUse active SF: enabled\n\t\tSpacing at SF min (%1): %2\n\t\tSpacing at SF max (%3): %4")
-								.arg(sfMin)
-								.arg(sfMinSpacing)
-								.arg(sfMax)
-								.arg(sfMaxSpacing));
-						}
-					}
-					else
-					{
-						//warn the user, not use Use Active SF and keep going
-						cmd.warning(QObject::tr("\tCan't use 'Use active SF': scalar field '%2' does not have any valid value.").arg(COMMAND_SET_ACTIVE_SF).arg(QString::fromStdString(sf->getName())));
-
+					if (params.spatial.tryEnableSfModulation(*sf)) {
+						cmd.print(QObject::tr("\tUse active SF: enabled\n\t\tSpacing at SF min (%1): %2\n\t\tSpacing at SF max (%3): %4")
+							.arg(params.spatial.sfMin)
+							.arg(params.spatial.sfMinSpacing)
+							.arg(params.spatial.sfMax)
+							.arg(params.spatial.sfMaxSpacing));
+					} else {
+						cmd.warning(QObject::tr("\tCan't use 'Use active SF' with scalar field '%2'").arg(QString::fromStdString(sf->getName())));
 					}
 				}
 			}
 
-			if (useActiveSF && !modParams.enabled)
+			if (useActiveSF && !params.spatial.modulationEnabled)
 			{
 				cmd.print(QObject::tr("\t'Use active SF' disabled. Falling back to constant spacing."));
 			}
 
-			CCCoreLib::ReferenceCloud* refCloud = CCCoreLib::CloudSamplingTools::resampleCloudSpatially(desc.pc, static_cast<PointCoordinateType>(step), modParams, nullptr, cmd.progressDialog());
-			if (!refCloud)
-			{
-				return cmd.error("Subsampling process failed!");
+			CCCoreLib::ReferenceCloud* refCloud = nullptr;
+			try {
+				refCloud = DoSubSample(params, desc.pc, cmd.progressDialog());
```

Comment:
Maybe you could catch the exception inside the DoSubSample and return nullptr. Because here it's not obvious by design that you'll get a non-null 'refCloud' pointer if no exception is raised. Which means I would still test refCloud against nullptr afterwards...

Reply:
If we keep exceptions from DoSubsample then the function should return a non pointer ReferenceCloud i think, otherwise if we remove exceptions then it should return a ptr that is null on error

Reply:
Yes, that's what most functions would do in CC, returning a nullptr in case of error